### DISCUSSION TITLE
The future of snakeyaml-engine and Kotlin Multiplatform

### DISCUSSION THREAD

**Discussion by krzema12** [krzema12/snakeyaml-engine-kmp#62] [General]
# Context

[Kotlin Multiplatform](https://kotlinlang.org/docs/multiplatform.html) (aka KMP, Kotlin MultiPlatform) is a tool where we can use Kotlin to target not only JVM, but also JavaScript and native platforms (iOS, WASM and more). YAML is very popular these days, so whoever is into Kotlin, there's demand to be able to use it outside of the JVM. For Kotlin, there's also a first-party (de)serialization library called [kotlinx.serialization](https://github.com/Kotlin/kotlinx.serialization/) - it doesn't implement any particular format, it's like a framework (a compiler plugin) for other libraries to consume and implement (de)serialization from/to concrete formats.

The most popular kotlinx.serialization-enabled library for YAML is [kaml](https://github.com/charleskorn/kaml) owned by @charleskorn, but it supports only the JVM because it depends on [snakeyaml-engine](https://bitbucket.org/snakeyaml/snakeyaml-engine/) which is a Java library. There's an alternative kotlinx.serialization-enabled library [yamlkt](https://github.com/Him188/yamlkt), but the (de)serialization logic is written there from scratch and it's missing some features that snakeyaml-engine already has.

Together with @aSemy, we've decided to give it a shot and bring proper YAML support for Kotlin.

See also:
* a discussion on Kotlin Slack: https://kotlinlang.slack.com/archives/C7A1U5PTM/p1684498259191979
* a related feature request for kaml to support Kotlin Multiplatform: https://github.com/charleskorn/kaml/issues/232

# Current state

Here's a repository where @aSemy and me have been working on a PoC to port snakeyaml-engine to Kotlin, and support targets beyond the JVM: https://github.com/krzema12/snakeyaml-engine-kmp/

To see a little demo of it already from within kaml, see https://github.com/krzema12/kaml/blob/implement-JS-target/demo-consumer/src/main/kotlin/Main.kt which prints:

```
====== Serialization ======
Serialized to YAML:
foo: "hey!"
bar: 123
baz:
  goo: 456.789
====== Deserialization ======
Deserialized from YAML:
SimpleClass(foo=deserialized-string, bar=987, baz=NestedClass(goo=11.22))
```

It proves that it's generally possible to port the library to Kotlin, and thanks to @aSemy's unbelievable engagement, it's been done within about 3 weeks. For sure there are certain edge cases that need attention, but the experiment shows that it's doable.

For now it's a completely separate project.

# Options

We'd like to discuss what's next, so that everyone is happy and it's good for the community.

## Port snakeyaml-engine to Kotlin and add support for Kotlin Multiplatform

It should be considered as the very first option, to avoid creating extra projects. It would involve working directly on https://bitbucket.org/snakeyaml/snakeyaml-engine/ to make the improvement. @asomov is the owner of the library. I'm sure the Kotlin community has lots of volunteers that would contribute to maintaining such improved library. It could also help with maintaining the library itself as a whole, not only the Kotlin or multiplarform aspect.

As we can read in https://groups.google.com/g/snakeyaml-core/c/MepazFf4ZwU from Andrey:

> For your information. I have re-written SnakeYAML in Kotlin 4 years ago. But it was not released because I do not see a clear advantage.
> What is the use case for Kotlin ?

It's probably right that considering the JVM alone, there's no much benefit in porting it to Kotlin, but 4 years ago Kotlin Multiplatform wasn't that big and popular.

Porting snakeyaml-engine to Kotlin would mean breaking changes for the JVM consumers because the JVM API leaks to the library API. Most prominently it's about `Optional` which in Kotlin is handled by compile-time nullability checks. The good news is that if we converted all `Optional`s to nullable references, it would still be usable to the Java consumers.

## Include snakeyaml-engine-kmp in kaml

Given that kaml is the most promising YAML solution for Kotlin, our separate port could be "adopted" into the library and maintained together as a whole, having proper separation between Gradle modules.

To be discussed in detail how it could look if we don't choose extending snakeyaml-engine.

## snakeyaml-engine-kmp as a separate project

We could stay on the current track, syncing from the parent repo (snakeyaml-engine) regularly.

To be discussed in detail how it could look if we don't choose extending snakeyaml-engine.

-----

Please let me know what you guys think, especially @asomov and @charleskorn!

**Comment by charleskorn** [comment]
For now, my suggestion is the third option (separate project) - this means you're not reliant on me in any way. Once things are a little more stable, then I'm happy to talk about the second option.

**Reply by krzema12** [reply]
Given the third option, I assume you would be open to depending on our library in non-JVM targets in kaml, assuming that the JVM target remains intact and the rest would be experimental? We'd publish it to Maven Central.

**Reply by charleskorn** [reply]
> Given the third option, I assume you would be open to depending on our library in non-JVM targets in kaml, assuming that the JVM target remains intact and the rest would be experimental? We'd publish it to Maven Central.

Yep, that's fine with me.

**Comment by asomov** [comment]
I do not quite catch what kind of feedback you expect from me. I will be glad to help and change SnakeYAML Engine when the changes do not crucially change the API. 
I am in favour to have only 1 deliverable per git repo. If you expect to create a JAR it should not be combineb with any other project.

**Reply by krzema12** [reply]
More specifically, we'd like to answer these questions:

- if you're fine with rewriting snakeyaml-engine to Kotlin. It wouldn't involve creating any extra JAR
  - rewriting here means rather "accepting contributions to rewrite it and maintain it as a Kotlin lib, with help to maintain it from the Kotlin community"
- if you're fine with supporting more platforms than just the JVM
  - if you perceive e. g. changing `Optionals` to nullable references a crucial API change

**Reply by asomov** [reply]
SnakeYAML Engine is already used in some major frameworks (like jackson).
I am open to create SnakeYAML Engine 3.0 with backwards incompatible changes (removing Optional etc)
Sorry, I do not get above message. I am fine with whatever to use SnakeYAML Engine in Kotlin. For me re-writing means - get rid of something in favour of something else. What is going to die ?

**Reply by aSemy** [reply]
@asomov SnakeYAML Engine 3.0 sounds like a good plan to me! Let's call it SYE3 for short.

Here are the issues I see that need to be resolved for an SYE3 release:

* Maintaining backwards compatibility for Java consumers
* updating SYE2 to make migrating easier
* Where & how to work

#### Backwards compatibility

Regarding maintaining backwards compatibility, I think it's achievable to maintain it without significant changes. There haven't been many changes to the ABI as far as Java consumers are concerned, as evidenced by the minimal changes to the existing Java tests. And even those minor changes can be re-worked to be smaller, or easy to migrate from.

The only other significant breaking changes I found while converting the Java to Kotlin Multiplatform are:

* Optional (as previously mentioned)
* Kotlin Multiplatform doesn't support reflection the same across all platforms. In particular, it doesn't support fetching the FQN of classes.
* Java IO (e.g. `InputStream`) replaced with [Okio](https://github.com/square/okio)

These classes and JVM specific functionality could be re-introduced in the Java specific `jvmMain` source code. However, in order to keep the codebase tidier it would be best to only use them in code specifically intended to be used by the library consumers.

#### Updating SYE2 to make migrating easier

Some work could be done to make migrating from SYE2 to SYE3 easier. I can only think of one example of this at the moment:

* Deprecate & replace functions that use Optional with equivalents that use `@Nullable` annotations

But there might be other improvements too...

#### Where & how to work

Where and how are we going to build SYE3?

Currently SYE2 is hosted on Bitbucket, which is missing a lot of features that other project-hosting sites have, so I'm not a fan. I'd rather work on GitHub, or an alternative like GitLab or Codeberg.

I propose that @asomov makes a new repo under https://github.com/snakeyaml, called `snakeyaml-engine-3` (or something), based on the [snakeyaml-engine-kmp](https://github.com/krzema12/snakeyaml-engine-kmp) fork. Add me and @krzema12 as contributors. And then we just need to set up publication of a snapshot version. 

What do you think @asomov @krzema12?

Alternatively we keep this fork alive, @krzema12 adds @asomov as a contributor, and we publish SYE3 to a new set of coordinates.

**Comment by asomov** [comment]
@aSemy SnakeYAML is in github for years - https://github.com/snakeyaml/snakeyaml-engine
Can you please show the places where  SnakeYAML _"support fetching the FQN of classes_." ? Or you mean the ones from the standard library ?
I do not get the statement about replacement with [Okio](https://github.com/square/okio). What is to be replaced ?

**Reply by aSemy** [reply]
> SnakeYAML is in github for years - https://github.com/snakeyaml/snakeyaml-engine

Great! So let's make a https://github.com/snakeyaml/snakeyaml-engine-3 repo, as I suggested above? Or do you want to use https://github.com/snakeyaml/snakeyaml-engine?

> Can you please show the places where SnakeYAML _"support fetching the FQN of classes."_? 

There were a few places that SYE uses `clazz.getName()`, but mostly in `toString()` methods. The only significant place is was in tags

https://github.com/snakeyaml/snakeyaml-engine/blob/5a35e1fe7f780d1405d5a03470f9f13d32b1638a/src/main/java/org/snakeyaml/engine/v2/nodes/Tag.java#L49-L57

I replaced this with the simple name in SnakeKMP

https://github.com/krzema12/snakeyaml-engine-kmp/blob/75f2e636e4c3244a77de4032ad3f50a430e5ea56/src/commonMain/kotlin/org/snakeyaml/engine/v2/nodes/Tag.kt#L40

(But I suspect this should be refactored so that `jvmMain` has a class-based function that will automatically use the class to get the FQN, and internally SYE3 uses a string-based tag. Using a string-based tag would better suite interopability with [Kaml](https://github.com/charleskorn/kaml), because KxS uses string-based tags too.)

> Or you mean the ones from the standard library ?

Yes, `clazz.getName()` was used to get the FQN of Optional and UUID I believe, but because SYE allows libraries to add their own types and tags other libraries could be adding them too.

> I do not get the statement about replacement with [Okio](https://github.com/square/okio). What is to be replaced ?

Kotlin Multiplatform doesn't support low-level byte operations very well, so Okio is needed instead of `java.io` equivalents 

* [YamlUnicodeReader](https://github.com/krzema12/snakeyaml-engine-kmp/blob/75f2e636e4c3244a77de4032ad3f50a430e5ea56/src/commonMain/kotlin/org/snakeyaml/engine/v2/api/YamlUnicodeReader.kt) for parsing non-UTF8 strings
* as a byte buffer
   https://github.com/krzema12/snakeyaml-engine-kmp/blob/75f2e636e4c3244a77de4032ad3f50a430e5ea56/src/commonMain/kotlin/org/snakeyaml/engine/v2/scanner/ScannerImpl.kt#L2127
* or as an input source 
   https://github.com/krzema12/snakeyaml-engine-kmp/blob/75f2e636e4c3244a77de4032ad3f50a430e5ea56/src/commonMain/kotlin/org/snakeyaml/engine/v2/scanner/StreamReader.kt#L37

and in other places, for similar reasons.

**Comment by asomov** [comment]
1. Why to make a repo ? Why not to use a branch ?
2. Of course I would like to keep using the repo !
3. The usage of FQN in Tag is redundant, it can be removed. It is a leftover from SnakeYAML
4. I do not understand how okio will work with JVM. Do we have to add a dependency ?

**Reply by aSemy** [reply]
1. A new repo is not necessary, it was just a suggestion. A branch is good too 👍. The git history is intact for snake-kmp so this should be easy. Do you want to create a new branch?
2. 🎉 
3. Gotcha, FQN can be removed 👍 
4. Yes, Okio would be a required dependency.
  
   It _might_ be possible to make sure that Okio is not used in any public-facing API (in Gradle terms it would be an `implementation` dependency, not an `api` dependency), but introduce a specific `org.snakeyaml:snake-yaml-engine-okio:3.0.0` dependency that adds additional Okio functionality. KxS does something similar and has [`json-okio`](https://github.com/Kotlin/kotlinx.serialization/tree/v1.5.1/formats/json-okio), and is helpful for non-JVM users.

   Using a tool like https://github.com/Kotlin/binary-compatibility-validator would be helpful for checking that Okio is not used in any public-facing API.

**Reply by aSemy** [reply]
On point 4: Just to clarify, there won't be _any_ need for JVM consumers to use Okio. They will still be able to use `java.io` types (e.g. input/output streams/readers), but SYE3 will convert those into Okio types for internal use, and convert Okio types back for public use.

The only point of relevance is that Okio is required dependency for Kotlin Multiplatform support, and there's no easy way around that.

**Reply by asomov** [reply]
@aSemy I am sorry I am confused with everything.
Why do you ask me to create a branch ? This is Open Source and this is github. You can fork and create any branch you want.
I would like to wait until I see PRs to have an overview of the coming changes.

**Reply by aSemy** [reply]
I'm sorry too, please could you clarify: when you say you are confused about everything, do you mean all of my answers? Or just specifically relating to the repository? If there are other things that are confusing, please ask 🙏 

We can make a new fork of https://github.com/snakeyaml/snakeyaml-engine, but a fork is a new repository (that happens to be linked to the original). Which is not a problem, but I want to make it clear.

I propose making a new branch in https://github.com/snakeyaml/snakeyaml-engine because there have been a _lot_ of changes to convert SYE to Kotlin Multiplatform. A branch would help because we can break down the changes into more specific PRs, which would make them easier for you to get an overview of the changes. That's what we did in this repo, so you can see an overview of the changes in the pull history, but these are the significant PRs:

* https://github.com/krzema12/snakeyaml-engine-kmp/pull/39 (this is a combined PR that updates the build config to use Gradle)
* https://github.com/krzema12/snakeyaml-engine-kmp/pull/40 (again, a combined PR that converts all `src/main/java` code to Kotlin/JVM)
* All PRs in https://github.com/krzema12/snakeyaml-engine-kmp/milestone/2?closed=1 converted the code to Kotlin Multiplatform

These PRs were incremental to make reviewing easier and to help make sure the Git history for each file was retained, even if it was converted from Java to Kotlin.

But maybe you want all the changes in one go? We can do that too?

**Reply by asomov** [reply]
I am confused with:
- what exactly needs to be done in SnakeYAML Engine
- how the changes should be done

> We can make a new fork of https://github.com/snakeyaml/snakeyaml-engine, but a fork is a new repository (that happens to be linked to the original). Which is not a problem, but I want to make it clear

This is how all the Open Source works. Make your fork and create PRs to contribute your improvements back to the parent.
- how do you work with other projects ?
- you do not expect to receive the full access to the parent project, do you ?

I do not understand why should I make a branch, because you can easily make it yourself in your fork and create a PR to add it to SnakeYAML. But I can add an empty branch - `kotlin-multiplatform`

**Comment by krzema12** [comment]
Regarding the repo, I'm confused. The GitHub repo is just a mirror, and the development happens on BitBucket. As far as I understood @aSemy, he'd like to develop on GitHub. Developing in a mirror repo sounds a bit awkward, but maybe it's doable. I was hoping we could move the development to GitHub and remove the BitBucket repo.

**Reply by asomov** [reply]
What is a mirror ? It is a fully functional project. I simply copy all the commits from bitbucket to github.
You may develop on github. What stops you ? 
Removing bitbucket is not an option. It does not harm.
There is obstacle to push your code to github.
